File: oscar/SetPartition.scala
/*******************************************************************************
 * OscaR is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 2.1 of the License, or
 * (at your option) any later version.
 *   
 * OscaR is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License  for more details.
 *   
 * You should have received a copy of the GNU Lesser General Public License along with OscaR.
 * If not, see http://www.gnu.org/licenses/lgpl-3.0.en.html
 ******************************************************************************/
package oscar.examples.cp.hakank

import oscar.cp.modeling._

import oscar.cp.core._
import scala.io.Source._
import scala.math._

/*

  Set partition problem in Oscar.

  Problem formulation from
  http://www.koalog.com/resources/samples/PartitionProblem.java.html
  """
  This is a partition problem.
  Given the set S = {1, 2, ..., n},
  it consists in finding two sets A and B such that:

   A U B = S,
   |A| = |B|,
   sum(A) = sum(B),
   sum_squares(A) = sum_squares(B)
  """

  Note: This model uses a binary matrix to represent the sets.
  

  @author Hakan Kjellerstrand hakank@gmail.com
  http://www.hakank.org/oscar/
 
*/

object SetPartition {

  def main(args: Array[String]) {

    val cp = CPSolver()

    //
    // data
    //
    val n        = if (args.length > 0) args(0).toInt else 16;
    val num_sets = if (args.length > 1) args(1).toInt else 2;

    val NRANGE = 0 until n
    val SRANGE = 0 until num_sets

    println("n: " + n + " num_sets: " + num_sets)

    //
    // variables
    // 
    // The matrix
    val a = Array.fill(num_sets,n)(CPIntVar(0 to 1)(cp))


    //
    // constraints
    //
    var numSols = 0

    cp.solve subjectTo {
    
      for(i <- SRANGE;
          j <- SRANGE if i!=j) {
          cp.add(
                 sum(for{k <- NRANGE} yield a(i)(k)*a(j)(k)) == 0
                 )
      }

      // ensure that all integers is in
      // (exactly) one partition
      cp.add(
             sum(
                 for{i <- SRANGE
                     j <- NRANGE} yield a(i)(j)
                 ) == n
             )

      

      for(i <- SRANGE; j <- SRANGE if i < j) {
        // same cardinality
        cp.add(
               sum(for{k <- NRANGE} yield a(i)(k) ) 
               ==
               sum(for{k <- NRANGE} yield a(j)(k) )
               )

        // same sum
        cp.add(
               sum(for{k <- NRANGE} yield a(i)(k)*k ) 
               ==
               sum(for{k <- NRANGE} yield a(j)(k)*k)
               )


        // same sum squared
        cp.add(
               sum(for{k <- NRANGE} yield a(i)(k)*k*a(i)(k)*k ) 
               ==
               sum(for{k <- NRANGE} yield a(j)(k)*k*a(j)(k)*k)
               )
      }

      // symmetry breaking for num_sets == 2
      if (num_sets == 2) {
        cp.add(a(0)(0) == 1)
      }


    } search {
       
      binaryStatic(a.flatten.toSeq)
    } onSolution {
      println("\nSolution:")
      var sums = 0
      var sums_squared = 0
      for(i <- SRANGE) {
        for(j <- NRANGE if a(i)(j).value == 1) {
          print((j+1) + " ")
          if (i == 0) {
            val v = (j+1)*a(i)(j).value
            sums += v
            sums_squared += v*v
          }
        }
        println()
      }

      println("Sums: " + sums  + " Sums squared: " + sums_squared)

      numSols += 1

    }
    println(cp.start())


  }

}

File: z3/set_partition.py
#!/usr/bin/python -u
# -*- coding: latin-1 -*-
# 
# Set partition problem in Z3
#
# Problem formulation from
# http://www.koalog.com/resources/samples/PartitionProblem.java.html
# '''
# This is a partition problem.
# Given the set S = {1, 2, ..., n},
# it consists in finding two sets A and B such that:
#    A U B = S,
#    |A| = |B|,
#    sum(A) = sum(B),
#    sum_squares(A) = sum_squares(B)
#
# '''
#
#  This model uses a binary matrix to represent the sets.

# 
# This Z3 model was written by Hakan Kjellerstrand (hakank@gmail.com)
# See also my Z3 page: http://hakank.org/z3/
# 
#
from __future__ import print_function
import sys
from z3_utils_hakank import *


#
# Partition the sets (binary matrix representation).
#
def partition_sets(sol, x, num_sets, n):
  for i in range(num_sets):
    for j in range(num_sets):
      if i != j:
        sol.add(Sum([x[i, k] * x[j, k] for k in range(n)]) == 0)

  # ensure that all integers is in
  # (exactly) one partition
  sol.add(Sum([x[i, j] for i in range(num_sets) for j in range(n)]) == n)


def main(n=16, num_sets=2):

  # Create the sol.
  sol = SolverFor("QF_FD")

  #
  # data
  #
  print("n:", n)
  print("num_sets:", num_sets)
  print()

  # Check sizes
  assert n % num_sets == 0, "Equal sets is not possible."

  #
  # variables
  #

  # the set
  a = {}
  for i in range(num_sets):
    for j in range(n):
      a[i, j] = makeIntVar(sol, "a[%i,%i]" % (i, j), 0, 1)

  a_flat = [a[i, j] for i in range(num_sets) for j in range(n)]

  #
  # constraints
  #

  # partition set
  partition_sets(sol, a, num_sets, n)

  for i in range(num_sets):
    for j in range(i, num_sets):

      # same cardinality
      sol.add(Sum([a[i, k] for k in range(n)])
              ==
              Sum([a[j, k] for k in range(n)]))

      # same sum
      sol.add(Sum([k * a[i, k] for k in range(n)])
              ==
              Sum([k * a[j, k] for k in range(n)]))

      # same sum squared
      sol.add(Sum([(k * a[i, k]) * (k * a[i, k])
                             for k in range(n)])
              ==
              Sum([(k * a[j, k]) * (k * a[j, k])
                             for k in range(n)]))

  # symmetry breaking for num_sets == 2
  if num_sets == 2:
    sol.add(a[0, 0] == 1)

  # search and result
  num_solutions = 0
  while sol.check() == sat:
    num_solutions += 1
    mod = sol.model()
    a_val = {}
    for i in range(num_sets):
      for j in range(n):
        a_val[i, j] = mod.eval(a[i, j]).as_long()

    sq = sum([(j + 1) * a_val[0, j] for j in range(n)])
    print("sums:", sq)
    sq2 = sum([((j + 1) * a_val[0, j]) ** 2 for j in range(n)])
    print("sums squared:", sq2)

    for i in range(num_sets):
      if sum([a_val[i, j] for j in range(n)]):
        print(i + 1, ":", end=' ')
        for j in range(n):
          if a_val[i, j] == 1:
            print(j + 1, end=' ')
        print()
    sol.add(Or([a[i,j] != mod.eval(a[i,j]) for i in range(num_sets) for j in range(n)]))
    print()

  print()
  print("num_solutions:", num_solutions)

n = 16
num_sets = 2
if __name__ == "__main__":
  if len(sys.argv) > 1:
    n = int(sys.argv[1])
  if len(sys.argv) > 2:
    num_sets = int(sys.argv[2])

  main(n, num_sets)
